File: bubble-sort.js
function bubbleSort(arr) {
var hasSwaped = true;
var temp;
  while (hasSwaped) {
    hasSwaped = false;
    for (var i = 0; i < arr.length - 1; i++) {
      if(arr[i] > arr[i+1]) {
        //use temp first
        temp = arr[i];
        //then use your arr[i] is free to assign.
        arr[i] = arr[i+1];
        arr[i+1] = temp;
        hasSwaped = true;
      }
    }
  }
  console.log(arr);
}

bubbleSort([5,1,4,2,8]);

// Scalability notation:
// O(n^2)

File: bubble_sort.rb
def bubble_sort(arr)
  n = arr.length
  loop do
    is_swapped = false
    (n-1).times do |i|
      if arr[i] > arr[i + 1]
        temp = arr[i]
        arr[i] = arr[i+1]
        arr[i+1] = temp
        is_swapped = true
      end
    end
    break if is_swapped == false
  end
  p arr
end

bubble_sort([100,34,345,1,-5,0])


# O(n^2) == bad